transports/quic/src/muxer.rs

```diff
@@ -79,24 +137,109 @@ impl QuicMuxer {
                 connection,
                 substreams: Default::default(),
-                pending_substreams: Default::default(),
+                poll_outbound_waker: None,
                 poll_close_waker: None,
-                poll_event_waker: None,
+                poll_inbound_waker: None,
             })),
         }
     }
 }
+impl StreamMuxer for QuicMuxer {
+    type Substream = Substream;
+    type Error = Error;
+
+    fn poll_address_change(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<libp2p_core::Multiaddr, Self::Error>> {
+        self.inner.lock().poll_connection(cx);
+        // TODO
+        Poll::Pending
+    }
+
+    fn poll_inbound(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>> {
+        let mut inner = self.inner.lock();
+        inner.poll_connection(cx);
+        let substream_id = match inner.connection.pop_incoming_substream() {
+        	Some(id) => id,
+        	None => {
+        		inner.poll_inbound_waker = Some(cx.waker().clone());
+        		return Poll::Pending;
+        	}
+        };
+        inner.substreams.insert(substream_id, Default::default());
+        let substream = Substream::new(substream_id, self.inner.clone());
+        Poll::Ready(Ok(substream))
+    }
+
+    fn poll_outbound(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Self::Substream, Self::Error>> {
+        let mut inner = self.inner.lock();
+        inner.poll_connection(cx);
+        let substream_id = match inner.connection.pop_outgoing_substream() {
+        	Some(id) => id,
+        	None => {
+        		inner.poll_outbound_waker = Some(cx.waker().clone());
+        		return Poll::Pending;
+        	}
+        };
+        inner.substreams.insert(substream_id, Default::default());
+        let substream = Substream::new(substream_id, self.inner.clone());
+        Poll::Ready(Ok(substream))
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let mut inner = self.inner.lock();
+        inner.poll_connection(cx);
+
+        if inner.connection.connection.is_drained() {
+            return Poll::Ready(Ok(()));
+        }
+
+        if inner.substreams.is_empty() {
+            let connection = &mut inner.connection;
+            if !connection.connection.is_closed() {
+                connection.close();
+                if let Some(waker) = inner.poll_inbound_waker.take() {
+                    waker.wake();
+                }
+            } else {
+            }
+            while let Poll::Ready(event) = inner.connection.poll_event(cx) {
+                if let ConnectionEvent::ConnectionLost(_) = event {
+                    return Poll::Ready(Ok(()));
+                }
+            }
+        } else {
+            for substream in inner.substreams.keys().cloned().collect::<Vec<_>>() {
```

Comment:
Unfortunately it's needed because otherwise the iterator still holds a reference to `inner`, for which we need a mutable reference within the loop. I'm not happy with it either; alternative solutions would be welcomed.

Reply:
This should do the trick:

```patch
Index: transports/quic/src/muxer.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/transports/quic/src/muxer.rs b/transports/quic/src/muxer.rs
--- a/transports/quic/src/muxer.rs	(revision b7103aa146621a6aa1504dc1a0b51d634883f9ae)
+++ b/transports/quic/src/muxer.rs	(date 1663725639634)
@@ -195,25 +195,30 @@
     }
 
     fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        let mut inner = self.inner.lock();
-        if inner.connection.connection.is_drained() {
+        let Inner {
+            substreams,
+            connection,
+            ..
+        } = &mut *self.inner.lock();
+
+        if connection.connection.is_drained() {
             return Poll::Ready(Ok(()));
         }
 
-        if inner.connection.connection.streams().send_streams() != 0 {
-            for substream in inner.substreams.keys().cloned().collect::<Vec<_>>() {
-                if let Err(e) = inner.connection.finish_substream(substream) {
+        if connection.connection.streams().send_streams() != 0 {
+            for substream in substreams.keys().copied() {
+                if let Err(e) = connection.finish_substream(substream) {
                     tracing::warn!("substream finish error on muxer close: {}", e);
                 }
             }
         }
         loop {
-            if inner.connection.connection.streams().send_streams() == 0
-                && !inner.connection.connection.is_closed()
+            if connection.connection.streams().send_streams() == 0
+                && !connection.connection.is_closed()
             {
-                inner.connection.close()
+                connection.close()
             }
-            match inner.connection.poll_event(cx) {
+            match connection.poll_event(cx) {
                 Poll::Ready(ConnectionEvent::ConnectionLost(_)) => return Poll::Ready(Ok(())),
                 Poll::Ready(_) => {}
                 Poll::Pending => break,

```

The key is, if you are borrowing different fields from the same struct, you just need to destruct them on the same statement so the borrow-checker gets that they are orthogonal :)